rascal-lsp/src/main/rascal/demo/lang/pico/LanguageServer.rsc

```diff
@@ -44,30 +44,44 @@ set[LanguageService] picoLanguageContributor() = {
 
 set[LanguageService] picoLanguageContributorSlowSummary() = {
     parser(parser(#start[Program])),
-    summarizer(picoSummarizer, providesImplementations = false)
+    analyzer(picoAnalyzer, providesImplementations = false),
+    builder(picoBuilder)
 };
 
 list[DocumentSymbol] picoOutliner(start[Program] input)
   = [symbol("<input.src>", DocumentSymbolKind::\file(), input.src, children=[
       *[symbol("<var.id>", \variable(), var.src) | /IdType var := input]
   ])];
 
-Summary picoSummarizer(loc l, start[Program] input) {
-    println("Running summary for pico!");
-    rel[str, loc] defs = {<"<var.id>", var.src> | /IdType var  := input};
+Summary picoAnalyzer(loc l, start[Program] input) = picoSummarizer(l, input, analyze());
+
+Summary picoBuilder(loc l, start[Program] input) = picoSummarizer(l, input, build());
+
+data PicoSummarizerMode = analyze() | build();
+
+Summary picoSummarizer(loc l, start[Program] input, PicoSummarizerMode mode) {
+    Summary s = summary(l);
+
+    rel[str, loc] defs = {<"<var.id>", var.src> | /IdType var := input};
     rel[loc, str] uses = {<id.src, "<id>"> | /Id id := input};
     rel[loc, str] docs = {<var.src, "*variable* <var>"> | /IdType var := input};
 
+    // Provide errors (cheap to compute) both in analyze mode and in build mode
+    s.messages += {<src, error("<id> is not defined", src)> | <src, id> <- uses, id notin defs<0>};
+    s.references += (uses o defs)<1,0>;
+    s.definitions += uses o defs;
+    s.documentation += (uses o defs) o docs;
 
-    return summary(l,
-        messages = {<src, error("<id> is not defined", src)> | <src, id> <- uses, id notin defs<0>},
-        references = (uses o defs)<1,0>,
-        definitions = uses o defs,
-        documentation = (uses o defs) o docs
-    );
+    // Provide warnings (expensive to compute) only in build mode
+    if (build() := mode) {
```

Comment:
`mode is build` is also nice

Reply:
why does `main` have two calls to registerLanguage?

Reply:
To test the feature if parallel language registries that both define a different part of the contributions.

Reply:
Ok, but this is demo code meant to educate newbies; so let's move that to test code and keep it simple here.

Reply:
I think this is now also used as tests, so we might have to split those two up. But let's to that outside this already big PR? (As this PR didn't introduce this split)

Reply:
Sure!